torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
   }
 #undef REQ
 
+  void fillTypeInfoMap(Block* block) {
+    for (Node* n : block->nodes()) {
+      if (n->kind() == prim::profile && n->outputs().size() == 1) {
+        if (auto tensor_ty = n->ty(attr::profiled_type)->cast<TensorType>()) {
+          GRAPH_DEBUG(
+              "TYPEINFO ADD: %", n->input()->debugName(), " --> ", *tensor_ty);
+          typeinfo_map_[n->input()] = tensor_ty;
+        }
+      }
+
+      for (Block* b : n->blocks()) {
+        fillTypeInfoMap(b);
+      }
+    }
+  }
+  void printTypeInfoMap() {
+    GRAPH_DEBUG("Typeinfo map:");
+    for (const auto& kv : typeinfo_map_) {
+      GRAPH_DEBUG("%", kv.first->debugName(), " --> ", *kv.second);
+    }
+  }
+
+  void removeProfilingNodes(Block* b) {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
+      if (it->kind() == prim::profile) {
+        for (auto o : it->outputs()) {
+          GRAPH_DEBUG("TYPEINFO ERASE: %", o->debugName());
+          typeinfo_map_.erase(o);
+        }
+        if (it->outputs().size()) {
+          it->output()->replaceAllUsesWith(it->input());
+        }
+        it.destroyCurrent();
+      } else {
+        for (Block* ib : it->blocks()) {
+          removeProfilingNodes(ib);
+        }
+      }
+    }
+  }
+  void guardFusionGroup(Node* n) {
+    auto subgraph = SubgraphUtils::getSubgraph(n);
+
+    // Fixup types of the subgraph inputs
+    std::vector<Value*> inputs_to_check;
+    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
+      // TODO: Make sure we use types encoded in the IR (not through the
+      // typeinfo map, but explicitly as value types)
+      if (typeinfo_map_.count(subgraph->inputs()[idx])) {
```

Comment:
Could you add a comment somewhere about how we only check the input types and use subgraph intermediary types for free because currently NNC only handles ops whose output types are uniquely determined by their inputs